sklearn/utils/extmath.py

```diff
@@ -117,29 +117,40 @@ def density(w, **kwargs):
 def safe_sparse_dot(a, b, dense_output=False):
     """Dot product that handle the sparse matrix case correctly
 
-    Uses BLAS GEMM as replacement for numpy.dot where possible
-    to avoid unnecessary copies.
-
     Parameters
     ----------
     a : array or sparse matrix
     b : array or sparse matrix
-    dense_output : boolean, default False
-        When False, either ``a`` or ``b`` being sparse will yield sparse
-        output. When True, output will always be an array.
+    dense_output : boolean, (default=False)
+        When False, ``a`` and ``b`` both being sparse will yield sparse output.
+        When True, output will always be a dense array.
 
     Returns
     -------
     dot_product : array or sparse matrix
-        sparse if ``a`` or ``b`` is sparse and ``dense_output=False``.
+        sparse if ``a`` and ``b`` are sparse and ``dense_output=False``.
     """
-    if sparse.issparse(a) or sparse.issparse(b):
-        ret = a * b
-        if dense_output and hasattr(ret, "toarray"):
-            ret = ret.toarray()
-        return ret
+    if a.ndim > 2 or b.ndim > 2:
```

Comment:
Does this really work for ND arrays? I would restrict it to 3D, which is what we are using, and raise a `NotImplementedError` otherwise. We are not testing 4D+ case anyway.

Reply:
Yes it does :)
I updated the tests to ensure that. I don't see any reason to raise an error since it's implemented and works.

Reply:
where do we need 3d?

Reply:
At least in coordinate descent (see diff) between 3D dense and sparse.